jax_cosmo/sparse.py

```diff
@@ -0,0 +1,361 @@
+
+
+@jit
+def sparse_dot_sparse(sparse1, sparse2):
+    """Calculate A @ B where A and B are both sparse and return sparse.
+
+    Inputs must be jax numpy arrays. No error checking is performed.
+    Use :func:`dot` for a more convenient front-end with error checking.
+
+    Parameters
+    ----------
+    sparse1 : array
+        3D array of shape (a, b, ndiag) of block diagonal elements.
+    sparse2 : array
+        3D array of shape (b, c, ndiag) of block diagonal elements.
+
+    Returns
+    -------
+    array
+        3D array of shape (a, c, ndiag) of block diagonal elements.
+    """
+    return vmap(
+        # Sparse multiply row @ col
+        vmap(
+            # Sparse multiply blocks B1 and B2
+            lambda B1, B2: np.sum(np.multiply(B1, B2), axis=0),
+            (0, None),
+            0,
+        ),
+        (None, 1),
+        1,
+    )(sparse1, sparse2)
+
+
+@jit
+def dense_dot_sparse_dot_dense(X, Y, Z):
+    """Calculate the bilinear form X @ Y @ Z where B is sparse.
+
+    Inputs must be jax numpy arrays. No error checking is performed.
+
+    Parameters
+    ----------
+    X : array
+        2D array of shape (a, b * ndiag) with dense matrix elements.
+    Y : array
+        3D array of shape (b, c, ndiag) with sparse matrix elements.
+    Z : array
+        2D array of shape (c * ndiag, d) with dense matrix elements.
+
+    Returns
+    -------
+    array
+        2D array of shape (a, d) with dense matrix elements.
+    """
+    return vmap(
+        vmap(
+            lambda row, sparse, col: np.dot(row, sparse_dot_vec(sparse, col)),
+            (None, None, 1),
+        ),
+        (0, None, None),
+    )(X, Y, Z)
+
+
+@jit
+def inv(sparse):
+    """Calculate the inverse of a square matrix in sparse format.
+
+    We currently assume that the matrix is invertible and you should not
+    trust the answer unless you know this is true (because jax.numpy.linalg.inv
+    has this behavior).
+
+    Parameters
+    ----------
+    sparse : array
+        3D array of shape (n, n, ndiag) of block diagonal elements.
+
+    Returns
+    -------
+    array
+        3D array of shape (n, n, ndiag) of block diagonal elements
+        representing the inverse matrix.
+    """
+    sparse = check_sparse(sparse, square=True)
+    return np.transpose(np.linalg.inv(np.transpose(sparse, (2, 0, 1))), (1, 2, 0))
+
+
+# We split the determinant calculation for a matrix with N x N blocks
+# into n pieces that can be evaluated in parallel, following eqn (2.2)
+# of https://arxiv.org/abs/1112.4379.  First build a helper function
+# to calculate one piece indexed by 0 <= k < N:
+@functools.partial(jit, static_argnums=(1, 2, 3))
+def _block_det(sparse, k, N, P):
+    u = sparse[k : k + 1, k + 1 : N, 0:P]
+    S = sparse[k + 1 : N, k + 1 : N, 0:P]
+    v = sparse[k + 1 : N, k : k + 1, 0:P]
+    Sinv_v = sparse_dot_sparse(inv(S), v)
+    return np.product(sparse[k, k] - sparse_dot_sparse(u, Sinv_v))
+
+
+@jit
+def det(sparse):
```

Comment:
For numerical stability I would recommend implementing the log determinant either instead or in addition. This usually is a lot more robust, especially in float32.